Added documentation for statistics service.

prasoon2211 opened the pull request:
This PR addresses the fourth todo item (to fix docs and add new docs) in the PR #1725 

@Stibbons @sa2ajj @benallard Please have a look.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -15,14 +15,15 @@ Stats Service
-   An instance of this class functions as a :class:`BuildbotService`. The instance of the running
-   service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more).
+   An instance of this class functions as a :class:`BuildbotService`.
+   The instance of the running service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more).
```
`... for more information).`

Stibbons commented:
+1

sa2ajj commented:
We do not seem to have any guidelines written regarding documentation formatting, so I'll put a few notes here and then try to put it somewhere "central":

* when documenting a class, we usually put `.. py:class:: full.import.name.to.Class`.
  This might lead to clashes when use in different places, so the secondary locations, should use `:noindex:` option
* while documenting parameters the best result is produced when they are documented as
```
``parameter``
    description
``another``
    and a descirption for this one
```

I am not sure whether `**The default callback**` really deserves that kind of formatting.  I'll think about during the day and return back in the evening...

prasoon2211 commented:
@sa2ajj I've made changes according to your comment.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -13,16 +13,19 @@ Here is a diagram demonstrating the working of the stats service:
-   * ``name``: The name of this service. This name can be used to access the running instance of this service using ``self.master.namedServices[name]``.
+   ``storage_backends``
+     A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
+   ``name``
```
do you think you could update the previously provided docs for consistency?

I just looked at the rendered docs and it does not look very nice.

prasoon2211 replied:
I did change the whole thing though (including the previously uploaded docs). I will see if anything remains that I haven't corrected.

prasoon2211 replied:
@sa2ajj Are you asking me to change :class:``StatsStorageBase`` to :class:``buildbot.path.to.StatsStorageBase``?

sa2ajj replied:
It's about `:param ...`, sorry for not being specific.

prasoon2211 replied:
Ah, okay.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -68,44 +70,68 @@ Stats Service
+.. py:class:: buildbot.statistis.storage_backends.base.StatsStorageBase
 
-   A base class for all storage services
+   A abstract base class for all storage services. It cannot be directly initialized - it would raise a ``TypeError`` otherwise.
```
`An abstract class`

Do not forget one-line-per-sentence rule.

prasoon2211 replied:
Missed this one. Will correct.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -68,44 +70,68 @@ Stats Service
-.. py:class:: InfluxStorageService
+   .. py:method:: thd_postStatsValue(self, post_data, series_name, context={})
+
+      :param post_data: A dict of key-value pairs that is sent for storage.
```
Let's be consistent about parameter documentation.

Though Sphinx offers this way to document parameters, it takes way too much space.

prasoon2211 replied:
Okay.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
there's no need for extra line here.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -13,16 +13,19 @@ Here is a diagram demonstrating the working of the stats service:
-   * ``storage_backends``: A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
-   * ``name``: The name of this service. This name can be used to access the running instance of this service using ``self.master.namedServices[name]``.
+   ``storage_backends``
+     A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
```
One sentence per line

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -68,44 +70,68 @@ Stats Service
+   A abstract base class for all storage services. It cannot be directly initialized - it would raise a ``TypeError`` otherwise.
 
-.. py:class:: InfluxStorageService
+   .. py:method:: thd_postStatsValue(self, post_data, series_name, context={})
```
actually, `{}` is dangerous as a default value.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
No need for this extra empty line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -68,44 +70,68 @@ Stats Service
 Storage backends
```
No need for an extra empty line before this one.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+       A dictionary of all build properties.
+     ``property_name``
+       Name of the build property to return.
+
```
No need for this extra line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CapturePropertyBase` for more information on this method.
+
```
No need for this extra line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      Since this class is used to capture a property over all builders, it simply returns ``True`` in all cases which means that all builders are allowed to post the property set
+      in ``property_name`` to storage backends.
```
One sentence per line (even if it's a long line).

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      Returns ``True`` for all builders.
+
```
No need for this extra line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CaptureBuildTimes` for more information on this method.
+
```
No need for this extra line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
No need for this extra line.

sa2ajj commented on master/docs/developer/stats-service.rst:
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CaptureBuildTimes` for more information on this method.
+
```
No need for this extra line.

sa2ajj commented on master/docs/manual/cfg-global.rst:
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
+   ``callback=None``
+     (Optional) A custom callback function for this class.
+     This callback function should take in two Python datetime objects - a ``start_time`` and an ``end_time`` and return a string that will be sent for storage in the storage backends.
+
```
No need for these two extra lines.

sa2ajj commented on master/docs/manual/cfg-global.rst:
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
+.. py:class:: buildbot.statistics.capture.CaptureDataAllBuilders
+   :noindex:
+
+   Instance of this capture class for capturing arbitrary data that is not stored as build-data on all builders
```
Missing `.` at the end of the sentence.

sa2ajj commented on master/docs/manual/cfg-global.rst:
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
 Along with using ``yieldMetricsValue``, the user will also need to use the ``CaptureData`` capture class.
+As an example, we can add the following to a build step:
+
+::
```
you might want to change it to

```
.. code-block:: python
```

we had quite a few issues when Sphinx/pygments was upgraded.

sa2ajj commented:
Looks good, though:

* we need consistent way of documenting parameters (at least, within the same class/file)
* there's no need for two or three (or more!) empty lines between "parts"

prasoon2211 commented:
@sa2ajj I've addressed the issues that were present in the last commit. Please review.

sa2ajj commented:
I built the docs locally and went through the rendered output.

Those "default callback" sections really confuse if they are not indented.

sa2ajj commented:
@prasoon2211 any updates?

prasoon2211 commented:
I've been a bit busy with school work. I'll work on your suggestions this evening.

prasoon2211 commented:
@sa2ajj So, I've fixed the problems and rebased on top of master. Also, I checked the rendered docs in browser and everything looked good to me. Please review.

sa2ajj commented:
Excellent.  Thank you
